File: app/models/Day.java
package models;

import com.github.jmkgreen.morphia.annotations.Embedded;
import com.github.jmkgreen.morphia.annotations.Entity;
import com.github.jmkgreen.morphia.annotations.Id;
import com.github.jmkgreen.morphia.annotations.Index;
import com.github.jmkgreen.morphia.annotations.Indexes;
import com.github.jmkgreen.morphia.annotations.PostLoad;
import com.github.jmkgreen.morphia.annotations.PrePersist;
import com.github.jmkgreen.morphia.annotations.Transient;
import com.google.common.collect.Lists;
import leodagdag.play2morphia.MorphiaPlugin;
import org.bson.types.ObjectId;
import org.codehaus.jackson.map.annotate.JsonDeserialize;
import org.joda.time.DateTime;
import utils.deserializer.DateTimeDeserializer;
import utils.time.TimerHelper;

import java.util.Date;
import java.util.List;

/**
 * @author f.patin
 */
@Entity
@Indexes({
	@Index("craId"),
	@Index("_date"),
	@Index("year, month")
})
public class Day {

	@Id
	public ObjectId id;

	public ObjectId craId;

	@Transient
	@JsonDeserialize(using = DateTimeDeserializer.class)
	public DateTime date;
	private Date _date;

	public Integer year;

	public Integer month;

	@Embedded
	public HalfDay morning;

	@Embedded
	public HalfDay afternoon;

	public String comment;

	public Boolean isSaturday() {
		return date != null && TimerHelper.isSaturday(date);
	}

	public Boolean isSunday() {
		return date != null && TimerHelper.isSunday(date);
	}

	public Boolean isDayOff() {
		return date != null && TimerHelper.isDayOff(date);
	}

	@SuppressWarnings({"unused"})
	@PrePersist
	private void prePersist() {
		_date = date != null ? date.toDate() : null;
		year = _date != null ? new DateTime(_date).getYear() : null;
		month = _date != null ? new DateTime(_date).getMonthOfYear() : null;
	}

	@SuppressWarnings({"unused"})
	@PostLoad
	private void postLoad() {
		date = _date != null ? new DateTime(_date.getTime()) : null;
	}

	public List<ObjectId> missionIds() {
		final List<ObjectId> result = Lists.newArrayList();
		if (morning != null) {
			result.addAll(morning.missionIds());
		}
		if (afternoon != null) {
			result.addAll(afternoon.missionIds());
		}
		return result;
	}

	public static List<Day> find(final ObjectId userId, final Integer year, final Integer month) {
		return MorphiaPlugin.ds().createQuery(Day.class)
			.field("craId").equal(userId)
			.field("year").equal(year)
			.field("month").equal(month)
			.asList();
	}
}

File: app/models/Day.scala
package models

import org.joda.time.DateTime
import play.api.Play.current
import play.api.libs.json.JsObject
import play.modules.reactivemongo.PlayBsonImplicits._
import play.modules.reactivemongo.ReactiveMongoPlugin
import reactivemongo.api.QueryBuilder
import reactivemongo.api.indexes.Index
import reactivemongo.bson.handlers.DefaultBSONHandlers.{DefaultBSONReaderHandler, DefaultBSONDocumentWriter}
import reactivemongo.bson.{BSONInteger, BSONDateTime, BSONDocument, BSONObjectID}
import reactivemongo.core.commands.LastError
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import reactivemongo.api.indexes.IndexType.Ascending


/**
 * @author f.patin
 */
/**
 * @author f.patin
 */
case class Day(id: Option[BSONObjectID],
               craId: BSONObjectID,
               _date: DateTime,
               year: Int,
               month: Int,
               morning: Option[HalfDay],
               afternoon: Option[HalfDay],
               comment: String) {
}

case class HalfDay(missionId: Option[BSONObjectID],
                   periods: Option[List[Period]]) {
	def isSpecial: Boolean = missionId.map(_ => true).getOrElse(false)
}

case class Period(missionId: BSONObjectID,
                  startTime: DateTime,
                  endTime: DateTime) {
}


object Day extends ToIndex {

	private val dbName = "Day"
	private val db = ReactiveMongoPlugin.db.collection(dbName)

	def ensureIndexes {
		import models.ensureIndex
		List(
			Index(List("craId" -> Ascending)),
			Index(List("year" -> Ascending, "month" -> Ascending)),
			Index(List("craId" -> Ascending, "_date" -> Ascending))
		).foreach {
			index => ensureIndex(index)(db)
		}
	}

	def fetch(userId: String, year: Int, month: Int, day: Int) = {
		val criterias = BSONDocument(
			"craId" -> BSONObjectID(userId),
			"_date" -> BSONDateTime(new DateTime(2013, 3, 1, 0, 0).getMillis)
		)
		implicit val reader = JsObjectReader
		val q = QueryBuilder().query(criterias)
		db.find[JsObject](q).headOption
	}

	def delete(userId: String, date: Long): Future[LastError] = {
		val s = BSONDocument(
			"craId" -> BSONObjectID(userId),
			"_date" -> BSONDateTime(date)
		)
		db.remove(s)
	}

	def delete(userId: String, date: Long, momentOfDay: String): Future[LastError] = {
		val s = BSONDocument(
			"craId" -> BSONObjectID(userId),
			"_date" -> BSONDateTime(date)
		)
		val u = BSONDocument("$unset" -> BSONDocument(momentOfDay -> BSONInteger(1)))
		db.update(s, u)
	}
}
